File: lib/core/time/clock.dart
// Polymorphism is intentional: SystemClock for prod, FakeClock for tests.
// ignore_for_file: one_member_abstracts, use_setters_to_change_properties

abstract class Clock {
  DateTime now();
}

class SystemClock implements Clock {
  const SystemClock();

  @override
  DateTime now() => DateTime.now();
}

class FakeClock implements Clock {
  FakeClock(this._now);

  DateTime _now;

  @override
  DateTime now() => _now;

  void advance(Duration delta) {
    _now = _now.add(delta);
  }

  void setTo(DateTime instant) {
    _now = instant;
  }
}

File: Sources/Core/Time/Clock.swift
import Foundation

// Injectable clock so domain logic is testable with a frozen / advanceable now.
// SystemClock in prod, FakeClock in tests. Ported from core/time/clock.dart.
protocol Clock {
  func now() -> Date
}

struct SystemClock: Clock {
  func now() -> Date { Date() }
}

final class FakeClock: Clock {
  private var current: Date

  init(_ now: Date) { current = now }

  func now() -> Date { current }

  func advance(_ delta: TimeInterval) { current = current.addingTimeInterval(delta) }

  func setTo(_ instant: Date) { current = instant }
}
